zebra-chain/src/primitives/zcash_history.rs

```diff
@@ -0,0 +1,194 @@
+    ///
+    /// `sapling_root` is the root of the Sapling note commitment tree of the block.
+    ///
+    /// Returns a vector of nodes added to the tree (leaf + internal nodes).
+    fn append_leaf(&mut self, block: &Block, sapling_root: &sapling::tree::Root) -> Vec<Node> {
+        let node_data = convert_block_to_librustzcash_data(block, self.network, sapling_root);
+        // TODO: handle error
+        let appended = self.tree.append_leaf(node_data).unwrap();
+
+        let mut new_nodes = Vec::new();
+        for entry in appended {
+            let mut node = Node {
+                node: [0; zcash_history::MAX_NODE_DATA_SIZE],
+            };
+            self.tree
+                .resolve_link(entry)
+                .expect("entry was just generated so it must be valid")
+                .data()
+                .write(&mut &mut node.node[..])
+                .expect("buffer was created with enough capacity");
+            new_nodes.push(node);
+        }
+        new_nodes
+    }
+
+    /// Append multiple blocks to the tree.
+    fn append_leaf_iter<'a>(
+        &mut self,
+        vals: impl Iterator<Item = (&'a Block, &'a sapling::tree::Root)>,
+    ) {
+        for (block, root) in vals {
+            self.append_leaf(block, root);
+        }
+    }
+
+    /// Remove the last leaf (block) from the tree.
+    ///
+    /// Returns the number of nodes removed from the tree after the operation.
+    fn truncate_leaf(&mut self) -> u32 {
+        // XXX: handle error
+        self.tree.truncate_leaf().unwrap()
+    }
+
+    /// Return the root hash of the tree, i.e. `hashChainHistoryRoot`.
+    fn hash(&self) -> [u8; 32] {
+        // Both append_leaf() and truncate_leaf() leave a root node, so it should
+        // always exist.
+        self.tree
+            .root_node()
+            .expect("must have root node")
+            .data()
+            .hash()
+    }
+}
+
+/// Convert a Block into a zcash_history::NodeData used in the MMR tree.
+///
+/// `sapling_root` is the root of the Sapling note commitment tree of the block.
+fn convert_block_to_librustzcash_data(
+    block: &Block,
+    network: Network,
+    sapling_root: &sapling::tree::Root,
+) -> zcash_history::NodeData {
+    // XXX: is this expect OK?
+    let height = block
+        .coinbase_height()
+        .expect("block must have coinbase height");
+    // XXX: is this expect OK?
+    let branch_id = ConsensusBranchId::current(network, height).expect("Must have branch ID");
+    let block_hash = block.hash().0;
+    let time: u32 = block
+        .header
+        .time
+        .timestamp()
+        .try_into()
+        .expect("deserialized and generated timestamps are u32 values");
+    let target = block.header.difficulty_threshold.0;
+    let sapling_root: [u8; 32] = (*sapling_root).into();
+    // XXX: is this expect OK?
+    let work = block
+        .header
+        .difficulty_threshold
+        .to_work()
+        .expect("must have work");
```

Comment:
Work is checked early in semantic verification, so it must be valid here:
```suggestion
    let work = block
        .header
        .difficulty_threshold
        .to_work()
        .expect("work must be valid during contextual verification");
```